src/base/utils.ml

```diff
@@ -1,48 +1,474 @@
+
+      let of_field x =
+        let open Let_syntax in
+        let%map () = assert_ (Constraint.boolean x) in
+        create x
+
+      module Unsafe = struct
+        let of_cvar (t : Cvar.t) : var = create t
+      end
+
+      module Assert = struct
+        let ( = ) (x : var) (y : var) = assert_equal (x :> Cvar.t) (y :> Cvar.t)
+
+        let is_true (v : var) = v = true_
+
+        let%snarkydef_ any (bs : var list) =
+          assert_non_zero (Cvar.sum (bs :> Cvar.t list))
+
+        let%snarkydef_ all (bs : var list) =
+          assert_equal
+            (Cvar.sum (bs :> Cvar.t list))
+            (Cvar.constant (Field.of_int (List.length bs)))
+
+        let%snarkydef_ exactly_one (bs : var list) =
+          assert_equal (Cvar.sum (bs :> Cvar.t list)) (Cvar.constant Field.one)
+      end
+
+      module Expr = struct
+        type t = Var of var | And of t list | Or of t list | Not of t
+
+        let rec eval t =
+          let open Let_syntax in
+          match t with
+          | Not t ->
+              eval t >>| not
+          | Var v ->
+              return v
+          | And ts ->
+              Checked.all (List.map ~f:eval ts) >>= all
+          | Or ts ->
+              Checked.all (List.map ~f:eval ts) >>= any
+
+        let assert_ t = eval t >>= Assert.is_true
+
+        let ( ! ) v = Var v
+
+        let ( && ) x y = And [ x; y ]
+
+        let ( &&& ) = ( && )
+
+        let ( || ) x y = Or [ x; y ]
+
+        let ( ||| ) = ( || )
+
+        let not t = Not t
+
+        let any xs = Or xs
+
+        let all xs = And xs
+      end
+    end
   end
 
-  let assert_equal ?label x y =
-    match (x, y) with
-    | Cvar0.Constant x, Cvar0.Constant y ->
-        if Field.equal x y then Checked.return ()
-        else
-          failwithf !"assert_equal: %{sexp: Field.t} != %{sexp: Field.t}" x y ()
-    | _ ->
-        Checked.assert_equal ?label x y
-
-  (* [equal_constraints z z_inv r] asserts that
-     if z = 0 then r = 1, or
-     if z <> 0 then r = 0 and z * z_inv = 1
-  *)
-  let equal_constraints (z : Cvar.t) (z_inv : Cvar.t) (r : Cvar.t) =
-    Checked.assert_all
-      [ Constraint.r1cs ~label:"equals_1" z_inv z Cvar.(constant Field.one - r)
-      ; Constraint.r1cs ~label:"equals_2" r z (Cvar.constant Field.zero)
-      ]
-
-  (* [equal_vars z] computes [(r, z_inv)] that satisfy the constraints in
-     [equal_constraints z z_inv r].
-
-     In particular, [r] is [1] if [z = 0] and [0] otherwise.
-  *)
-  let equal_vars (z : Cvar.t) : (Field.t * Field.t) As_prover.t =
-    let open As_prover in
-    let%map z = As_prover.read_var z in
-    if Field.equal z Field.zero then (Field.one, Field.zero)
-    else (Field.zero, Field.inv z)
+  module Boolean = Private.Boolean
+
+  let equal, mul, square, div, inv = Private.(equal, mul, square, div, inv)
+
+  let if_, assert_non_zero = Private.(if_, assert_non_zero)
+
+  let equal_vars, equal_constraints = Private.(equal_vars, equal_constraints)
```

Comment:
I'm not sure what the aim here is. 
Every function of this module is public for there is no module signature constraining what's exposed by the `Make` functor. 

If we want function aliases defined at toplevel, we could basically do `include Private` and we'd have the same aliases (and way more). 

If we want to limit what's available within the API, we'd like to do it either though defining a module type, or by defining the module interface (the sibling `.mli` file, which does not currently exist)

Reply:
I just wanted to limit while working, I will make an .mli